packages/@uppy/core/src/Uppy.js

```diff
@@ -400,194 +371,73 @@ class Uppy {
-        }
-
-        // otherwise this is likely an extension
-        if (type[0] === '.' && file.extension) {
-          return file.extension.toLowerCase() === type.substr(1).toLowerCase()
-        }
-        return false
-      })
-
-      if (!isCorrectFileType) {
-        const allowedFileTypesString = allowedFileTypes.join(', ')
-        throw new RestrictionError(this.i18n('youCanOnlyUploadFileTypes', { types: allowedFileTypesString }))
-      }
-    }
-
-    // We can't check maxTotalFileSize if the size is unknown.
-    if (maxTotalFileSize && file.size != null) {
-      let totalFilesSize = 0
-      totalFilesSize += file.size
-      files.forEach((f) => {
-        totalFilesSize += f.size
-      })
-      if (totalFilesSize > maxTotalFileSize) {
-        throw new RestrictionError(this.i18n('exceedsSize', {
-          size: prettierBytes(maxTotalFileSize),
-          file: file.name,
-        }))
-      }
-    }
-
-    // We can't check maxFileSize if the size is unknown.
-    if (maxFileSize && file.size != null) {
-      if (file.size > maxFileSize) {
-        throw new RestrictionError(this.i18n('exceedsSize', {
-          size: prettierBytes(maxFileSize),
-          file: file.name,
-        }))
-      }
-    }
-
-    // We can't check minFileSize if the size is unknown.
-    if (minFileSize && file.size != null) {
-      if (file.size < minFileSize) {
-        throw new RestrictionError(this.i18n('inferiorSize', {
-          size: prettierBytes(minFileSize),
-        }))
-      }
+  /*
+  * @constructs
+  * @param { Error } error
+  * @param { undefined } fileOrFiles
+  */
+  /*
+  * @constructs
+  * @param { RestrictionError } error
+  * @param { UppyFile } fileOrFiles
+  */
+  /*
+  * @constructs
+  * @param { RestrictionError } error
+  * @param { Array<UppyFile> } fileOrFiles
+  */
+  #informAndEmit (error, fileOrFiles) {
+    const { message, details = '' } = error
+
+    if (error.isRestriction) {
+      this.emit('restriction-failed', fileOrFiles, error)
+    } else {
+      this.emit('error', error)
     }
+    this.info({ message, details }, 'error', this.opts.infoTimeout)
+    this.log(`${message} ${details}`.trim(), 'error')
   }
 
-  /**
-   * Check if minNumberOfFiles restriction is reached before uploading.
-   *
-   * @private
-   */
-  #checkMinNumberOfFiles (files) {
-    const { minNumberOfFiles } = this.opts.restrictions
-    if (Object.keys(files).length < minNumberOfFiles) {
-      throw new RestrictionError(`${this.i18n('youHaveToAtLeastSelectX', { smart_count: minNumberOfFiles })}`)
+  validateRestrictions (file, files = this.getFiles()) {
+    // TODO: directly return the Restriction error in next major version.
+    // we create RestrictionError's just to discard immediately, which doesn't make sense.
+    try {
+      this.#restricter.validate(file, files)
+      return { result: true }
+    } catch (err) {
+      return { result: false, reason: err.message }
```

Comment:
Shouldn't we send the whole error object? What if the `err` is nullish?
```suggestion
    } catch (reason) {
      return { result: false, reason }
```